### PR TITLE
Add API for querying the state after a list of events

### PR DESCRIPTION
**Pull request by NegativeMjark**
This will be used to query whether the server has enough context to process an event received over federation.

### CODE REVIEW

File: src/github.com/matrix-org/dendrite/roomserver/api/query.go [hunk 1]
```diff
@@ -44,18 +44,51 @@ type QueryLatestEventsAndStateResponse struct {
...
+type QueryStateAfterEventsResponse struct {
+	// Copy of the request for debugging.
+	QueryStateAfterEventsRequest
+	// Does the room exist on this roomserver?
+	// If the room doesn't exist this will be false and StateEvents will be empty.
+	RoomExists bool
+	// Do all the previous events exist on this roomserver?
+	// If some of previous events do not exist this will be false and StateEvents will be empty.
+	PrevEventsExist bool
+	// The state events requested.
```
Comment (Kegsay) [inline src/github.com/matrix-org/dendrite/roomserver/api/query.go]: Any ordering?

File: src/github.com/matrix-org/dendrite/roomserver/query/query.go [hunk 2]
```diff
@@ -101,6 +149,20 @@ func (r *RoomserverQueryAPI) SetupHTTP(servMux *http.ServeMux) {
 			return util.JSONResponse{Code: 200, JSON: &response}
 		}),
 	)
+	servMux.Handle(
+		api.RoomserverQueryStateAfterEventsPath,
+		makeAPI("query_state_after_events", func(req *http.Request) util.JSONResponse {
```
Comment (Kegsay) [inline src/github.com/matrix-org/dendrite/roomserver/query/query.go]: Swap to `common` please.

File: src/github.com/matrix-org/dendrite/roomserver/state/state.go [hunk 3]
```diff
@@ -316,6 +316,76 @@ func loadStateAtSnapshotForNumericTuples(
...
+	if len(prevStates) == 1 {
+		// Fast path for a single event.
+		prevState := prevStates[0]
+		result, err := loadStateAtSnapshotForNumericTuples(
+			db, prevState.BeforeStateSnapshotNID, stateKeyTuples,
+		)
+		if err != nil {
+			return nil, err
+		}
+		if prevState.IsStateEvent() {
```
Comment (Kegsay) [inline src/github.com/matrix-org/dendrite/roomserver/state/state.go]: I don't understand this. Surely `prevState` is always state and its an error if it isn't?

Why do we loop `result` if it is state and clobber elements, but don't clobber elements if it isn't state?

File: src/github.com/matrix-org/dendrite/roomserver/state/state.go [hunk 4]
```diff
@@ -316,6 +316,76 @@ func loadStateAtSnapshotForNumericTuples(
...
+				}
+			}
+		}
+		return result, nil
+	}
+
+	// Slow path for more that one event.
+	// Load the entire state so that we can do conflict resolution if we need to.
+	// TODO: The are some optimistations we could do here:
+	//    1) We only need to do conflict resolution if the is a conflict in the
```
Comment (Kegsay) [inline src/github.com/matrix-org/dendrite/roomserver/state/state.go]: > if the is a

File: src/github.com/matrix-org/dendrite/roomserver/state/state.go [hunk 5]
```diff
@@ -316,6 +316,76 @@ func loadStateAtSnapshotForNumericTuples(
...
+	//       requested tuples so we might try loading just those tuples and then
+	//       checking for conflicts.
+	//    2) When there is a conflict we still only need to load the state
+	//       needed to do conflict resolution which would save us having to load
+	//		 the full state.
```
Comment (Kegsay) [inline src/github.com/matrix-org/dendrite/roomserver/state/state.go]: Spurious tab.

File: src/github.com/matrix-org/dendrite/roomserver/query/query.go (hunk 2 above)
Comment (NegativeMjark) [inline src/github.com/matrix-org/dendrite/roomserver/query/query.go]: Done 8a42fa6e

File: src/github.com/matrix-org/dendrite/roomserver/state/state.go (hunk 3 above)
Comment (NegativeMjark) [inline src/github.com/matrix-org/dendrite/roomserver/state/state.go]: Commented 1baa6fe4

File: src/github.com/matrix-org/dendrite/roomserver/state/state.go (hunk 4 above)
Comment (NegativeMjark) [inline src/github.com/matrix-org/dendrite/roomserver/state/state.go]: Fixed 1baa6fe4

File: src/github.com/matrix-org/dendrite/roomserver/state/state.go (hunk 5 above)
Comment (NegativeMjark) [inline src/github.com/matrix-org/dendrite/roomserver/state/state.go]: Fixed 1baa6fe4

File: src/github.com/matrix-org/dendrite/roomserver/api/query.go (hunk 1 above)
Comment (NegativeMjark) [inline src/github.com/matrix-org/dendrite/roomserver/api/query.go]: None. Same as QueryLatestEventsAndState.

**Comment by Kegsay** [comment]
LGTM!

**Pull request merged**